src/vm/eventtrace.cpp

```diff
@@ -3722,6 +3722,75 @@ VOID ETW::TypeSystemLog::DeleteTypeHashNoLock(AllLoggedTypes **ppAllLoggedTypes)
 
 //---------------------------------------------------------------------------------------
 //
+// Called from shutdown to give us the opportunity to dump any sampled object allocation
+// information before the process shuts down.
+//
+
+// static
+VOID ETW::TypeSystemLog::FlushObjectAllocationEvents()
+{
+    LIMITED_METHOD_CONTRACT;
+
+    // If logging is not enabled, then we don't need to do any work.
+    if (!(s_fHeapAllocHighEventEnabledNow || s_fHeapAllocHighEventEnabledNow))
+    {
+        return;
+    }
+
+    AllLoggedTypes * pThreadAllLoggedTypes = NULL;
+    Thread * pThread = NULL;
+
+    // Get the thread store lock.
+    ThreadStoreLockHolder tsl;
+
+    // Iterate over each thread and log any un-logged allocations.
+    while ((pThread = ThreadStore::GetThreadList(pThread)) != NULL)
+    {
+        pThreadAllLoggedTypes = pThread->GetAllocationSamplingTable();
+        if (pThreadAllLoggedTypes == NULL)
+        {
+            continue;
+        }
+
+        DWORD dwAllocsSkippedForSample;
+        SIZE_T cbIgnoredSizeForSample;
+
+        // Iterate over each module.
+        AllLoggedTypesHash * pLoggedTypesHash = &pThreadAllLoggedTypes->allLoggedTypesHash;
```

Comment:
@Jkotas, this was actually one of the things I wanted to get some confirmation on from you.

I put the call to this function after we finish finalization and if necessary, suspend threads and trap threads, so that there shouldn’t be any more managed allocations taking place.  Managed allocations and full blocking GCs are the only thing that can manipulate the list, and to my knowledge GCs can’t happen at this point.  So, as long as managed allocations can’t happen either, we can walk these lists safely.

From: Jan Kotas [mailto:notifications@github.com]
Sent: Friday, March 13, 2015 9:38 PM
To: dotnet/coreclr
Cc: Brian Robbins
Subject: Re: [coreclr] Flush sampled allocations to event stream at runtime shutdown (#440)


In src/vm/eventtrace.cpp<https://github.com/dotnet/coreclr/pull/440#discussion_r26435311>:

> +    ThreadStoreLockHolder tsl;

> +

> +    // Iterate over each thread and log any un-logged allocations.

> +    while ((pThread = ThreadStore::GetThreadList(pThread)) != NULL)

> +    {

> +        pThreadAllLoggedTypes = pThread->GetAllocationSamplingTable();

> +        if (pThreadAllLoggedTypes == NULL)

> +        {

> +            continue;

> +        }

> +

> +        DWORD dwAllocsSkippedForSample;

> +        SIZE_T cbIgnoredSizeForSample;

> +

> +        // Iterate over each module.

> +        AllLoggedTypesHash * pLoggedTypesHash = &pThreadAllLoggedTypes->allLoggedTypesHash;

What guarantees that no other threads are modifying these lists? Maybe worth a comment.

—
Reply to this email directly or view it on GitHub<https://github.com/dotnet/coreclr/pull/440/files#r26435311>.